### ISSUE TITLE
Existing resource conflict

### ISSUE THREAD

**Issue by autarchprinceps** [fluxcd/helm-operator#249]
**Describe the bug**

When migrating from the old flux helm chart with integrated operator to this, we are getting issues with some, but definitively not all charts, that it fails to initially upgrade the releases, because an update would create resources that already exist (in my point of view, that's the point of an update, but maybe I'm misunderstanding something). This message is displayed in the kubectl get hr MESSAGE column. As one can see below, a lot of different resource kinds are affected.
```
kube-system             external-dns                   external-dns                   deployed   failed to upgrade chart for release [external-dns]: rendered manifests contain a new resource that already exists. Unable to continue with update: existing resource conflict: kind: Deployment, namespace: kube-system, name: external-dns                                       19m
kube-system             k8s-spot-termination-handler   k8s-spot-termination-handler   deployed   failed to upgrade chart for release [k8s-spot-termination-handler]: rendered manifests contain a new resource that already exists. Unable to continue with update: existing resource conflict: kind: ServiceAccount, namespace: kube-system, name: k8s-spot-termination-handler   19m
kube-system             kube2iam                       kube2iam                       deployed   failed to upgrade chart for release [kube2iam]: rendered manifests contain a new resource that already exists. Unable to continue with update: existing resource conflict: kind: ClusterRole, namespace: , name: kube2iam    
```

**To Reproduce**

We updated it along the lines of the method mentioned here: https://github.com/fluxcd/helm-operator/issues/215

**Expected behavior**

Either it should work this way already, or at least we would like some kind of force override flag, to allow for a smooth in place upgrade. Completely deleting and recreating stuff like kube2iam isn't really an option in production.

**Logs**

Logs contain very similar message, as kubectl get hr above. It looks like this repeatedly:
```
ts=2020-01-27T10:11:30.562027707Z caller=helm.go:66 component=helm version=v3 info="preparing upgrade for kube2iam"
ts=2020-01-27T10:11:30.777875611Z caller=helm.go:66 component=helm version=v3 info="performing update for cluster-autoscaler"
ts=2020-01-27T10:11:30.831786466Z caller=helm.go:66 component=helm version=v3 info="dry run for cluster-autoscaler"
ts=2020-01-27T10:11:31.050821351Z caller=helm.go:66 component=helm version=v3 info="performing update for kube2iam"
ts=2020-01-27T10:11:31.092699535Z caller=release.go:185 component=release release=kube2iam targetNamespace=kube-system resource=kube-system:helmrelease/kube2iam helmVersion=v3 error="failed to determine if the release should be synced" err="failed to upgrade chart for release [kube2iam]: rendered manifests contain a new resource that already exists. Unable to continue with update: existing resource conflict: kind: ClusterRole, namespace: , name: kube2iam"
```

**Additional Context**

Running 1.0.0-rc8 of Helm Operator. Issue is independent from Flux itself, since HelmRelease Objects are created correctly, but not applied correctly.

**Comment by stefanprodan** [comment]
@autarchprinceps there is an option to force the upgrade:
```yaml
kind: HelmRelease
spec:
  forceUpgrade: true
```

Can you give that a try please?

**Comment by hiddeco** [comment]
@stefanprodan this is not going to help without us passing the flag to Helm when performing the dry-run.

**Comment by hiddeco** [comment]
@autarchprinceps can you please confirm setting the above mentioned flag in combination with `hiddeco/helm-operator:bug-dry-run-force-0cc46b00` solves this for you?

**Comment by autarchprinceps** [comment]
When adding this the message remains for some, other releases get new errors. For example traefik gets the following error, despite the fact that the real helm values or chart version and therefore the result of the templates shouldn't have changed at all
```failed to upgrade chart for release : failed to replace object: Service "traefik-dashboard" is invalid: spec.clusterIP: Invalid value: "": field is immutable```

**Comment by hiddeco** [comment]
@autarchprinceps what happens when you add `spec.resetValues` to the release?

Providing the flags that would make a `helm upgrade` command succeed would help resolving this by the way.

**Comment by autarchprinceps** [comment]
@hiddeco I don't see any difference. The messages with resetValues look the same as with just forceUpgrade. I tried deploying it with both at the same time in a second, newly migrated cluster as well.

As for the helm commands that would resolve this, as of now this is truely resolvable only by deleting the helm release that was created prior to the helm operator and letting the helm operator install it.
It still has the same values then, but somehow it is treated differently, when later upgrades are made to a release handled by the operator, or at least we hope so, since we haven't had a case yet, where a change causes this to break upgrade in a running flux/helmv3 cluster.
We haven't had the same messages with manual helm upgrades either.

**Comment by hiddeco** [comment]
> It still has the same values then, but somehow it is treated differently, when later upgrades are made to a release handled by the operator, or at least we hope so, since we haven't had a case yet, where a change causes this to break upgrade in a running flux/helmv3 cluster.

Deleting the `HelmRelease` deletes the release (from Helm's storage), it is thus treated as a new installation without the migrated storage included. This is a sign for me that Helm can't cope with the migrated storage, as the error is also returned by Helm and not an error from us.

> We haven't had the same messages with manual helm upgrades either.

I find it highly unlikely that this does not happen for releases the operator has issues with as it literally performs the same actions as the Helm binary, but using their published Go packages. The `HelmRelease` with `.spec.forceUpgrade` set is nothing more than: `helm upgrade -n <target namespace> -i <release name> -f <merged defined values> --force <chart source>`.

**Comment by derrickburns** [comment]
fyi https://github.com/helm/helm/issues/6646

**Comment by derrickburns** [comment]
and https://github.com/helm/helm/issues/6850

This is probably the bug fix: https://github.com/jason-liew/helm/commit/739d735b818da2d719dcdad0a72ea4f6076c50c7

**Comment by runningman84** [comment]
I have a similar problem but it shows "failed to compose values for chart release" in the message:
```
╰─ kubectl get hr --all-namespaces                           
NAMESPACE               NAME                           RELEASE                        STATUS     MESSAGE                                      AGE
botkube                 botkube                        botkube                        deployed   Helm release sync succeeded                  7d19h
cert-manager            cert-manager                   cert-manager                   deployed   failed to compose values for chart release   7d19h
monitoring              prometheus-operator            prometheus                     deployed   failed to compose values for chart release   7d19h
````
These are the corresponding logs:
```
I0128 08:45:09.814502       8 upgrade.go:79] preparing upgrade for cert-manager
ts=2020-01-28T08:45:09.974095234Z caller=release.go:188 component=release release=prometheus targetNamespace=monitoring resource=monitoring:helmrelease/prometheus-operator helmVersion=v3 error="failed to determine if the release should be synced" err="failed to upgrade chart for release [prometheus]: rendered manifests contain a new resource that already exists. Unable to continue with update: existing resource conflict: kind: PodSecurityPolicy, namespace: , name: prometheus-grafana"
I0128 08:45:10.154508       8 upgrade.go:87] performing update for cert-manager
ts=2020-01-28T08:45:10.59661369Z caller=release.go:188 component=release release=cert-manager targetNamespace=cert-manager resource=cert-manager:helmrelease/cert-manager helmVersion=v3 error="failed to determine if the release should be synced" err="failed to upgrade chart for release [cert-manager]: rendered manifests contain a new resource that already exists. Unable to continue with update: existing resource conflict: kind: Deployment, namespace: cert-manager, name: cert-manager-cainjector"
```

**Comment by autarchprinceps** [comment]
I have tried it to verify it, it is indeed not just an issue in the initial step. It does occur sometimes in later updates as well.

**Comment by hiddeco** [comment]
> I have a similar problem but it shows "failed to compose values for chart release" in the message

@runningman84 you are likely running a version prior to `1.0.0-rc8`, as this was fixed in #209.

> I have tried it to verify it, it is indeed not just an issue in the initial step. It does occur sometimes in later updates as well.

Given the issues mentioned by @derrickburns, I can now conclude (instead of speculate) this is not an issue with the operator but with Helm itself, and until this has been fixed in the upstream there is little we can do about it. I will thus be closing this to get it of our list, but feel free to discuss any workarounds here.
